// @ks89/angular-modal-gallery/esm2022/lib/components/modal-gallery/attach-to-overlay.service.mjs
import { Injectable, Injector } from '@angular/core';
import { ModalGalleryRef } from './modal-gallery-ref';
import { ModalGalleryComponent } from './modal-gallery.component';
import { DIALOG_DATA } from './modal-gallery.tokens';
import { ComponentPortal } from '@angular/cdk/portal';
import * as i0 from "@angular/core";
import * as i1 from "./modal-gallery.service";
class AttachToOverlayService {
    constructor(injector, modalGalleryService) {
        this.injector = injector;
        this.modalGalleryService = modalGalleryService;
    }
    /**
     * To be called by a provider with the APP_INITIALIZER token.
     */
    initialize() {
        this.modalGalleryService.triggerAttachToOverlay.subscribe(payload => this.attachToOverlay(payload));
    }
    /**
     * Private method to attach ModalGalleryComponent to the overlay.
     * @param payload {@link AttachToOverlayPayload} with all necessary information
     * @private
     */
    attachToOverlay(payload) {
        const injector = Injector.create({
            parent: this.injector,
            providers: [
                { provide: ModalGalleryRef, useValue: payload.dialogRef },
                { provide: DIALOG_DATA, useValue: payload.config }
            ]
        });
        const containerPortal = new ComponentPortal(ModalGalleryComponent, null, injector);
        payload.overlayRef.attach(containerPortal);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: AttachToOverlayService, deps: [{ token: i0.Injector }, { token: i1.ModalGalleryService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: AttachToOverlayService, providedIn: 'root' }); }
}
export { AttachToOverlayService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: AttachToOverlayService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i0.Injector }, { type: i1.ModalGalleryService }]; } });
