// @ks89/angular-modal-gallery/esm2022/lib/modal-gallery.module.mjs
/*
 The MIT License (MIT)

 Copyright (C) 2017-2023 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import { COMPONENTS, CarouselComponent } from './components/components';
import { PlainGalleryComponent } from './components/plain-gallery/plain-gallery.component';
import { DIRECTIVES } from './directives/directives';
import { AttachToOverlayService } from './components/modal-gallery/attach-to-overlay.service';
import * as i0 from "@angular/core";
import * as i1 from "./components/plain-gallery/plain-gallery.component";
import * as i2 from "./components/carousel/carousel.component";
import * as i3 from "./components/carousel/carousel-previews/carousel-previews.component";
import * as i4 from "./components/upper-buttons/upper-buttons.component";
import * as i5 from "./components/dots/dots.component";
import * as i6 from "./components/previews/previews.component";
import * as i7 from "./components/current-image/current-image.component";
import * as i8 from "./components/current-image/loading-spinner/loading-spinner.component";
import * as i9 from "./components/accessible.component";
import * as i10 from "./components/modal-gallery/modal-gallery.component";
import * as i11 from "./directives/click-outside.directive";
import * as i12 from "./directives/size.directive";
import * as i13 from "./directives/keyboard-navigation.directive";
import * as i14 from "./directives/wrap.directive";
import * as i15 from "./directives/direction.directive";
import * as i16 from "./directives/a-tag-bg-image.directive";
import * as i17 from "./directives/description.directive";
import * as i18 from "./directives/margin.directive";
import * as i19 from "./directives/max-size.directive";
import * as i20 from "./directives/fallback-image.directive";
import * as i21 from "./directives/swipe.directive";
/**
 * Module to import it in the root module of your application.
 */
class GalleryModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: GalleryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.0", ngImport: i0, type: GalleryModule, declarations: [i1.PlainGalleryComponent, i2.CarouselComponent, i3.CarouselPreviewsComponent, i4.UpperButtonsComponent, i5.DotsComponent, i6.PreviewsComponent, i7.CurrentImageComponent, i8.LoadingSpinnerComponent, i9.AccessibleComponent, i10.ModalGalleryComponent, i11.ClickOutsideDirective, i12.SizeDirective, i13.KeyboardNavigationDirective, i14.WrapDirective, i15.DirectionDirective, i16.ATagBgImageDirective, i17.DescriptionDirective, i18.MarginDirective, i19.MaxSizeDirective, i20.FallbackImageDirective, i21.SwipeDirective], imports: [CommonModule, OverlayModule], exports: [PlainGalleryComponent, CarouselComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: GalleryModule, providers: [
            {
                provide: APP_INITIALIZER,
                multi: true,
                deps: [AttachToOverlayService],
                useFactory: (service) => {
                    return () => service.initialize();
                }
            }
        ], imports: [CommonModule, OverlayModule] }); }
}
export { GalleryModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: GalleryModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, OverlayModule],
                    declarations: [COMPONENTS, DIRECTIVES],
                    providers: [
                        {
                            provide: APP_INITIALIZER,
                            multi: true,
                            deps: [AttachToOverlayService],
                            useFactory: (service) => {
                                return () => service.initialize();
                            }
                        }
                    ],
                    exports: [PlainGalleryComponent, CarouselComponent]
                }]
        }] });
